// rainbow-bridge-lib/src/get-config.js

function getConfig(env) {
    switch (env) {
        case 'mainnet':
            return {
                networkId: 'mainnet',
                nearNodeUrl: 'https://rpc.mainnet.near.org',
                ethNodeUrl: '',
            };
        case 'testnet':
            return {
                networkId: 'default',
                nearNodeUrl: 'https://rpc.testnet.near.org',
                ethNodeUrl: '',
            };
        case 'local':
        case 'test':
            return {
                networkId: 'local',
                // NEAR configuration.
                nearNodeUrl: 'http://localhost:3030',
                keyPath: `${process.env.HOME}/.near/local/validator_key.json`,
                masterAccount: 'test.near',
                nearEthClientId: 'client.test.near',
                nearEthProverId: 'prover.test.near',
                // Ethereum configuration.
                ethNodeUrl: 'http://localhost:9545',
                ethFromSecretKey: '0x4910a61dba22bf489e9b8fb94cadb1418b6c41ce2ef47b695ea2d153b36e6aee',
                ethProverAddress: '',
            };
        default:
            throw new Error(`Unconfigured environment '${env}'. Can be configured in src/config.js.`);
    }
}

module.exports = getConfig;
