// tez-bridge-lib/src/get-config.js

function getConfig(env) {
    switch (env) {
        case 'mainnet':
            return {
                networkId: 'mainnet',
                tezosNodeUrl: 'https://rpc.mainnet.tezos.org',
                ethNodeUrl: '',
            };
        case 'testnet':
            return {
                networkId: 'default',
                tezosNodeUrl: 'https://rpc.testnet.tezos.org',
                ethNodeUrl: '',
            };
        case 'local':
        case 'test':
            return {
                networkId: 'local',
                // TEZOS configuration.
                tezosNodeUrl: 'http://localhost:3030',
                keyPath: `${process.env.HOME}/.tezos/local/validator_key.json`,
                masterAccount: 'test.tezos',
                tezosEthClientId: 'client.test.tezos',
                tezosEthProverId: 'prover.test.tezos',
                // Ethereum configuration.
                ethNodeUrl: 'http://localhost:9545',
                ethFromSecretKey: '0x4910a61dba22bf489e9b8fb94cadb1418b6c41ce2ef47b695ea2d153b36e6aee',
                ethProverAddress: '',
            };
        default:
            throw new Error(`Unconfigured environment '${env}'. Can be configured in src/config.js.`);
    }
}

module.exports = getConfig;
