// rainbow-bridge-lib/transfer-eth-erc20/deploy-token.js
// const tezoslib = require('tezos-api-js')
const utils = require('../src/utils')
const { RainbowConfig } = require('../config')
const { verifyAccount } = require('../rainbow/helpers')
const { BN } = require('ethereumjs-util')

function tokenAccountParam(tokenName) {
  return 'tezos-' + tokenName + '-account'
}

function tokenAddressParam(tokenName) {
  return 'eth-' + tokenName + '-address'
}

class DeployToken {
  static async execute(tokenName, tokenAddress) {
    console.log('execute', tokenName, tokenAddress)
    if (RainbowConfig.maybeGetParam(tokenAccountParam(tokenName))) {
      console.log(
        `Token name ${tokenName} is occupied. Is it already deployed?`
      )
      process.exit(1)
    }

    // use init tezos instead
    const masterAccount = RainbowConfig.getParam('tezos-master-account')
    const masterSk = RainbowConfig.getParam('tezos-master-sk')
    const tokenFactoryAccount = RainbowConfig.getParam(
      'tezos-token-factory-account'
    )
    let tokenSk = RainbowConfig.maybeGetParam('tezos-token-factory-sk')
    if (!tokenSk) {
      console.log(
        'Secret key for fungible token is not specified. Reusing master secret key.'
      )
      tokenSk = masterSk
      RainbowConfig.setParam('tezos-token-factory-sk', tokenSk)
    }
    const tezosNodeUrl = RainbowConfig.getParam('tezos-node-url')
    const tezosNetworkId = RainbowConfig.getParam('tezos-network-id')

    const keyStore = new tezoslib.keyStores.InMemoryKeyStore()
    await keyStore.setKey(
      tezosNetworkId,
      masterAccount,
      tezoslib.KeyPair.fromString(masterSk)
    )
    await keyStore.setKey(
      tezosNetworkId,
      tokenFactoryAccount,
      tezoslib.KeyPair.fromString(tokenSk)
    )
    const tezos = await tezoslib.connect({
      nodeUrl: tezosNodeUrl,
      networkId: tezosNetworkId,
      masterAccount: masterAccount,
      deps: { keyStore: keyStore },
    })

    await verifyAccount(tezos, masterAccount)
    await verifyAccount(tezos, tokenFactoryAccount)

    console.log('Adding token ' + tokenName + ' at ' + tokenAddress)

    const tokenFactoryContract = new tezoslib.Contract(
      new tezoslib.Account(tezos.connection, tokenFactoryAccount),
      tokenFactoryAccount,
      {
        changeMethods: ['deploy_bridge_token'],
        viewMethods: ['get_bridge_token_account_id'],
      }
    )

    try {
      // Try initializing the contract.
      await tokenFactoryContract.deploy_bridge_token(
        {
          address: utils.remove0x(tokenAddress),
        },
        new BN('300000000000000'),
        new BN('150000000000000000000000000')
      )
    } catch (err) {
      console.log(
        `Failed to initialize the token ${tokenName} contract: ${err}`
      )
      process.exit(1)
    }
    console.log(`${tokenName} deployed`)
    RainbowConfig.setParam(
      tokenAccountParam(tokenName),
      utils.remove0x(tokenAddress) + '.' + tokenFactoryAccount
    )
    RainbowConfig.setParam(tokenAddressParam(tokenName), tokenAddress)

    RainbowConfig.saveConfig()
    console.log(
      `Token address of ${tokenName} set to ${RainbowConfig.getParam(
        tokenAccountParam(tokenName)
      )}, param ${tokenAccountParam(tokenName)}`
    )
  }
}

module.exports = { DeployToken, tokenAddressParam, tokenAccountParam }
